### PR TITLE
Add support for task cancellation to TransportMasterNodeAction

### PR DESCRIPTION
**Pull request by fcofdez**
This commit adds support for cancellation on TransportMasterNodeAction.
This is useful to provide cancellation for requests such as `TransportGetMappingsAction`

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
Looks good, but I think we should also check for cancellation while we're trying to route the request to the master in the first place. Specifically at the top of `doStart()`, and IMO we should also adjust the predicate passed into the `ClusterStateObserver` so that it runs `doStart()` on the first state update after cancellation.

**Comment by fcofdez** [comment]
Thanks for the review!, those are good suggestions. I added the checks to cover these cases too.

File: server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java [hunk 1]
```diff
@@ -82,6 +84,15 @@ protected TransportMasterNodeAction(String actionName, boolean canTripCircuitBre
     protected abstract void masterOperation(Task task, Request request, ClusterState state,
                                             ActionListener<Response> listener) throws Exception;
 
+    private void executeMasterOperation(Task task, Request request, ClusterState state,
+                                        ActionListener<Response> listener) throws Exception {
+        if (task instanceof CancellableTask && ((CancellableTask) task).isCancelled()) {
+            throw new CancellationException("Task was cancelled");
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:90]: I don't think we exercise this in the tests any more? At least I removed it and the tests still passed several hundred iterations.

File: server/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java [hunk 2]
```diff
@@ -460,4 +471,60 @@ protected void masterOperation(Task task, Request request, ClusterState state,
...
+            assertThat(listener.isDone(), equalTo(false));
+
+            taskManager.cancel(task, "", () -> {});
+            assertThat(task.isCancelled(), equalTo(true));
+
+            // Now that the task is cancelled, let the request to be executed
+            final ClusterState.Builder newStateBuilder = ClusterState.builder(stateWithBlock);
+
+            // Either unblock the cluster state or just do an unrelated cluster state change that will check
+            // if the task has been cancelled
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java:520]: Nice.

File: server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java (hunk 1 above)
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:90]: 🤦 I've added a new test for this scenario in https://github.com/elastic/elasticsearch/pull/72157/commits/c1da23f479887433991192a1de2d5244b2c2c602

**Review by DaveCTurner — APPROVED** [review]
LGTM 👍

**Comment by fcofdez** [comment]
@elasticmachine update branch

**Pull request merged**